memory/src/main/scala/filodb.memory/data/OffheapLFSortedIDMap.scala

```diff
@@ -246,22 +286,173 @@ class OffheapLFSortedIDMapReader(memFactory: MemFactory, holderKlass: Class[_ <:
     innerBinSearch(0, mapLen)
   }
 
+  /**
+   * Acquire exclusive access to this map, spinning if necessary. Exclusive lock isn't re-entrant.
+   */
+  def acquireExclusive(inst: MapHolder): Unit = {
+    // Spin-lock implementation. Because the owner of the shared lock might be blocked by this
+    // thread as it waits for an exclusive lock, deadlock is possible. To mitigate this problem,
+    // timeout and retry, allowing shared lock waiters to make progress. The timeout doubles
+    // for each retry, up to a limit, but the retries continue indefinitely.
+
+    var timeoutNanos = 1.millisecond.toNanos
+    var warned = false
+
+    while (true) {
+      if (tryAcquireExclusive(inst, timeoutNanos)) {
```

Comment:
Wonder if it would make sense to have some metrics to measure and report contention (like wait times). This can help us correlate slowness in ingestion to contention in block acquisition.

Reply:
Most effective strategy would be to rely on JMX or something and sample all the threads for hotspots.